giraph-core/src/main/java/org/apache/giraph/counters/CustomCounters.java

```diff
@@ -0,0 +1,124 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.giraph.counters;
+
+import java.util.ArrayList;
+import java.util.HashSet;
+import java.util.List;
+import java.util.Set;
+
+/**
+ * Custom counters created by the client or app
+ * Stores the names of the counters, and the final aggregated values
+ * Since there is no way to get the list of available counters from the context,
+ * we need to call addCustomCounter to store the information about the list of
+ * available counters. This is accessed by the worker/master to fetch the
+ * final counter values based on the group and counter name.
+ * It is assumed that whenever a counter is created, addCustomCounter is called,
+ * to enable aggregations on that counter, using the Giraph mechanism for
+ * aggregating counters.
+ */
+public class CustomCounters {
```

Comment:
The data structures here are modified and accessed by multiple threads, they will need to be synchronized appropriately. Maybe this is why you get a concurrentmodification as per your comment in getCustomerCounters()?